README.md
## [Мой Круг в порядке](https://chrome.google.com/webstore/detail/%D0%BC%D0%BE%D0%B9-%D0%BA%D1%80%D1%83%D0%B3-%D0%B2-%D0%BF%D0%BE%D1%80%D1%8F%D0%B4%D0%BA%D0%B5/mdoeebdabadnmfejoipdeaemfedepanl)

###### Расширение для Chrome, которое сортирует вакансии и заказы на сайтах [Мой Круг](https://moikrug.ru) и [Фрилансим](https://freelansim.ru) по дате

После установки расширения самые свежие вакансии и заказы всегда будут первыми в списке, как и должно быть на любом нормальном сайте по поиску работы.

У создателей Моего Круга и Фрилансим на этот счет было особое мнение и базовую функцию сортировки они решили не предоставлять.

Расширение работает даже если список вакансий динамически обновился через AJAX (поиск, фильтрация), благодаря [DOM Mutation Observers API](https://www.w3.org/TR/dom/#mutation-observers)

[![Chrome Web Store](https://developer.chrome.com/webstore/images/ChromeWebStore_BadgeWBorder_v2_340x96.png)](https://chrome.google.com/webstore/detail/%D0%BC%D0%BE%D0%B9-%D0%BA%D1%80%D1%83%D0%B3-%D0%B2-%D0%BF%D0%BE%D1%80%D1%8F%D0%B4%D0%BA%D0%B5/mdoeebdabadnmfejoipdeaemfedepanl)


CHANGELOG.md
## Changelog

### 2.1
- Refactor and get rid of self-made helper functions with lodash
- Optimize performance by reducing the amount of event listeners
- Remove unused background scripts
- Add EN and RU locales

### 2.0
- Add [freelansim.ru](https://freelansim.ru) support
- Remade small sized icons (16px & 19px) to be more visible

### 1.2
- Refactor with ES6 language spec
- 100% valid ESLint code
- Updated jQuery to version 3 and using slim version without AJAX and animations

### 1.1
- Dynamical updates via DOM Mutation Observers
- Modern icons instead of defaults

### 1.0
- Initial release


src/inject/inject.js
function curried(fn, ...args) {
  return (...nArgs) => fn.apply(this, [...args, ...nArgs]);
}

function replaceAll(str, map) {
  const re = new RegExp(Object.keys(map).join('|'), 'gi');
  return str.replace(re, matched => map[matched]);
}

function parseRusMonth(str) {
  const months = {
    января: 'jan',
    февраля: 'feb',
    марта: 'mar',
    апреля: 'apr',
    мая: 'may',
    июня: 'jun',
    июля: 'jul',
    августа: 'aug',
    сентября: 'sep',
    октября: 'oct',
    ноября: 'nov',
    декабря: 'dec',
  };
  return Date.parse(replaceAll(str, months));
}

function parseRusDate(str) {
  const num = str.match(/\d+/)[0];
  const mod = str.match(/(мин|час|дн|ден|мес|год)/)[0];
  const calendar = {
    мин: 60,
    час: 60 * 60,
    дн: 60 * 60 * 24,
    ден: 60 * 60 * 24,
    мес: 60 * 60 * 24 * 30, // I know, I know, sorry
    год: 60 * 60 * 24 * 365,
  };
  const millis = num * calendar[mod] * 1000;
  return new Date(Date.now() - millis);
}

function observeNode(nodeId, performSorting) {
  let latestURI;
  $(nodeId).observe('childlist added', (record) => {
    if (record.addedNodes.length > 0) {
      if (latestURI !== record.addedNodes[0].baseURI) performSorting();
      latestURI = record.addedNodes[0].baseURI;
    }
  });
}

function sortJobList(nodeId, childClass, dateClass, parser) {
  const node = $(nodeId);
  const childList = node.children(childClass);

  childList.sort((a, b) => {
    const an = parser($(a).find(dateClass).html());
    const bn = parser($(b).find(dateClass).html());

    if (an > bn) return -1;
    if (an < bn) return 1;
    return 0;
  });

  childList.detach().appendTo(node);
}

$(document).ready(() => {
  let nodeId;
  let performSorting;

  if (location.host === 'moikrug.ru') {
    nodeId = '#jobs_list';
    performSorting = curried(sortJobList, nodeId, '.job', 'span.date', parseRusMonth);
    $('#vacancies_search_suggest_form_q').observe({ attributes: true, attributeFilter: ['class'] }, performSorting);
  }

  if (location.host === 'freelansim.ru') {
    nodeId = '#tasks_list';
    performSorting = curried(sortJobList, nodeId, '.content-list__item', 'span.params__published-at', parseRusDate);
  }

  performSorting();
  observeNode(nodeId, performSorting);
});


src/sortjobs.js
const sortJobList = (nodeId, childClass, dateClass, titleClass, dateParser, addedRecords) => {
  if (!addedRecords) return;

  const added = Array.from(addedRecords.addedNodes).filter(
    (el) => $(el).hasClass(childClass)
  );
  const sorted = _.orderBy(
    added,
    [
      (el) => dateParser($(el).find(dateClass).html()),
      (el) => $(el).find(titleClass).attr('title')
    ],
    'desc'
  );

  if (isEqualArrayOfNodes(added, sorted)) return;

  $(nodeId).empty();
  $(sorted).appendTo(nodeId);
}

const parseRusMonth = (str) => {
  const rusMonths = [
    'января',  'февраля', 'марта',
    'апреля',  'мая',     'июня',
    'июля',    'августа', 'сентября',
    'октября', 'ноября',  'декабря'
  ];
  const engMonths = [
    'jan', 'feb', 'mar',
    'apr', 'may', 'jun',
    'jul', 'aug', 'sep',
    'oct', 'nov', 'dec'
  ];
  const months = _.zipObject(rusMonths, engMonths);
  return Date.parse(
    _.replace(
      str,
      new RegExp(rusMonths.join('|'), 'gi'),
      (match) => months[match]
    )
  );
}

const parseRusDate = (str) => {
  const num = str.match(/\d+/)[0];
  const mod = str.match(/(мин|час|дн|ден|мес|год)/)[0];
  const calendar = {
    мин: 60,
    час: 60 * 60,
    дн:  60 * 60 * 24,
    ден: 60 * 60 * 24,
    мес: 60 * 60 * 24 * 30, // Variable month length doesn't really matter here
    год: 60 * 60 * 24 * 365,
  };
  const millis = num * calendar[mod] * 1000;
  return new Date(Date.now() - millis);
}

const isEqualArrayOfNodes = (right, left) => {
  if (right.length !== left.length) return false;
  for (let i = 0; i < right.length; i++) {
    if (!right[i].isEqualNode(left[i])) {
      return false;
    }
  }
  return true;
}

$(document).ready(() => {
  let nodeId;
  let performSorting;

  if (location.host === 'moikrug.ru') {
    nodeId = '#jobs_list';
    performSorting = _.partial(
      sortJobList,
      nodeId,
      'job',
      'span.date',
      '.title',
      parseRusMonth
    );
  }

  if (location.host === 'freelansim.ru') {
    nodeId = '#tasks_list';
    performSorting = _.partial(
      sortJobList,
      nodeId,
      'content-list__item',
      'span.params__published-at',
      '.task__title',
      parseRusDate
    );
  }

  // Initial sorting
  performSorting({
    addedNodes: $(nodeId).children()
  });
  // Sorting after AJAX
  $(nodeId).observe(
    'added',
    (record) => performSorting(record)
  );
});
